ISSUE-199 - Grouping of icons to search for an owner/co-owner

LPoin opened the pull request:
!! MERGE AFTER [PR 202](https://github.com/georchestra/mapstore2-cadastrapp/pull/202) FOR ISSUE-197 !!

- Add a new icon in the toolbar for searching owner & co-owner
- Modify positon of search owner & co-owner buttons
- Search owner & co-owner panel unchanged
- Modify TButton to accept custom icons

LPoin committed:
Modify Cadastrapp button icon

LPoin committed:
Add USERS search tool and group owner and co-owners

LPoin committed:
remove cadastrapp icon to create a specific branch for issue 175

landryb commented:
same as #203, needs plumbing to install the svg, otherwise you get a broken icon:
![image](https://github.com/user-attachments/assets/b529782c-f8db-4cc7-a593-2cc9d638692d)

remark on the selected tool icon: it should be **green** not grey to hilight which one from owner/co-owner is active.. same thing for the new button color, it should be green if we're on the owner tools pane.


there's a weird behaviour, if you click on the (already selected) tool button, it .. comes back to the cadastrapp welcome pane ?

pierrejego commented:
Not tested in integration for the moment, only in dev

LPoin committed:
fix custom icon svg build import

LPoin committed:
lint

Gaetanbrl commented:
> needs plumbing to install the svg

We find build patch to includes SVG files in build process. I dunno if they finally works well....

https://github.com/georchestra/mapstore2-cadastrapp/blob/e2f87a8c015090541ac5a87052da5bd4cf1a4710/build/extension/prod-webpack.config.js#L31-L39

For this pull, we can set image source from svg data string.

Gaetanbrl commented on js/extension/plugins/cadastrapp/toolbar/SearchTools.jsx:
```diff
@@ -39,24 +45,29 @@ const BUTTONS_SETTINGS = {
+                    [SEARCH_TOOLS.OWNER, SEARCH_TOOLS.COOWNER].includes(k) :
+                    [SEARCH_TOOLS.PLOT, SEARCH_TOOLS.OWNERS].includes(k)
+                )
+                // .filter(k => {
```
Code mort à supprimer ?

LPoin replied:
Non commenté temporairement pour afficher les outils.
C'est remis en place.

Gaetanbrl commented on js/extension/plugins/cadastrapp/toolbar/SearchTools.jsx:
```diff
@@ -39,24 +45,29 @@ const BUTTONS_SETTINGS = {
-                        bsStyle={isActive && "active"}
+                        bsStyle={isActive && "success"}
                         {...BUTTONS_SETTINGS[toolName]}
+                        isCustom = {toolName === "OWNERS"}
```
Je ne suis pas pour cette modification car elle ne concerne que le bouton "OWNERS".
Si on souhaite utiliser un autre SVG pour un autre bouton, alors ce code ne serait plus bon et serait à modifier à nouveau.

Je propose d'ajouter une propriété `imgSrc` dans les boutons, je vais commenter les endroits en ce sens.

Gaetanbrl commented on js/extension/plugins/cadastrapp/toolbar/TButton.jsx:
```diff
@@ -12,14 +12,15 @@ export default ({
             bsStyle={bsStyle || "primary"}
             className="square-button">
-            <Glyphicon glyph={glyph} />
+            {isCustom ? glyph : <Glyphicon glyph={glyph} />}
```
Je serai plus pour une utilisation d'une path et d'une classe car le glyph devrait être réservé à une valeur `string` de glyph icon.

Exemple :

```
{{props.imgSrc ? 
    <img src={props.imgSrc} className={props.cls} />
    : <Glyphicon glyph={glyph} />
}}
```

Avec ce code, n'importe quel bouton qui a une propriété `imgSrc` pourrait utiliser une source.

LPoin committed:
change owners icon colors

Gaetanbrl commented on js/extension/plugins/cadastrapp/toolbar/SearchTools.jsx:
```diff
@@ -7,19 +7,25 @@ import TButton from './TButton';
         tooltip: tooltip("search", "cadastrapp.parcelle.tooltip")
     },
+    [SEARCH_TOOLS.OWNERS]: {
+        glyph: <img src={ svgDataUrl } className="ownersIcon"/>,
```
La valeur de la propriété glyph devrait être réservée à une clé glyphicon.

Exemple en complément de mon autre commentaire sur le fichier TButton :

```
    [SEARCH_TOOLS.PLOT]: {
        glyph: "", // or no glyph key
        imgSrc: svgDataUrl,
        cls: "ownersIcon",
         tooltip: tooltip("users", "cadastrapp.rechercheProprietaires.tooltip")
    },
```

Gaetanbrl commented on js/extension/plugins/cadastrapp/SearchSection.jsx:
```diff
@@ -60,12 +61,25 @@ export default function SearchSection({
-        /><OwnersList /></>);
+        return (
+            <>
+                <div className="selectionToolsButton">
```
Class à renommer de préférence avec un prefix type cadapp (au choix)

Gaetanbrl reviewed (commented):
J'opterai pour une autre manière de gérer l'icône avec une image en n'utilsant pas uniquement la clé **glyph** (pour moi réservée dans MapStore à une valeur de glyphicon).

LPoin commented:
Tested in integration.
Mapstore 2023.02

- SVG fixed
- "if you click on the (already selected) tool button, it .. comes back to the cadastrapp welcome pane", it's the current behaviour, if you click on en tool already selected, it close the panel and return to welcome panel.

LPoin committed:
modify TButton custom and fix class name

LPoin committed:
uncomment code

LPoin committed:
fix selection tool by default

landryb commented:
> For this pull, we can set image source from [svg data string](https://github.com/georchestra/mapstore2-cadastrapp/pull/205/commits/19b223d036b0cd74910a5a57aa4fc0698d620308#diff-0c00974814750a3f948d253618b94fa2cd330670ebf06dc576ea5d5d28e39f94R18).

im not a fan of this solution since it makes it very hard to (eventually) override the icon.. we should be able to use svgs in extensions, im surprised there's no working example elsewhere ?

i dunno if its my local build, but it seems this PR/ranch includes #202 since i get to test both when building this branch... which is perfect from a testing POV, not sure from a project mgmt POV :)

the ppl search icon renders okay here, there's just a small inconsistency in title positioning wrt icons in the pane (above/below):

![image](https://github.com/user-attachments/assets/7d8ff265-854f-4b3e-a8a2-464c40e1625e)
vs
![image](https://github.com/user-attachments/assets/5b5095ff-d134-423f-b0d0-cc337e64e06c)

that makes the display weird when you switch from one to the other.

> * "if you click on the (already selected) tool button, it .. comes back to the cadastrapp welcome pane", it's the current behaviour, if you click on en tool already selected, it close the panel and return to welcome panel.

ah, maybe that's the current behaviour, right. But ... from an UX pov, isn't it... super confusing ? :)

Gaetanbrl commented:
> im surprised there's no working example elsewhere ?

Same for me. But SVG's build config seems only apply in mapstore2-cadatrapp repo.

This works well in dev mode but not after extension build process (due to different build config / process).

> it makes it very hard to (eventually) override the icon

You mean if you want to override via the extensions directory (datadir) ?

Gaetanbrl commented:
> (due to different build config / process).

I suppose webpack config on simple `npm:start` use mapstore2 config with SVG management :

- https://github.com/georchestra/mapstore2-cadastrapp/blob/master/package.json#L112
- https://github.com/georchestra/mapstore2-cadastrapp/blob/master/webpack.config.js

...and extension  build (to create zip) use a different config :
- https://github.com/georchestra/mapstore2-cadastrapp/blob/master/package.json#L113

landryb commented:
> You mean if you want to override via the extensions directory (datadir) ?

yes, since here the extension is unpacked in the georchestra datadir, having the icon as an asset/svg file makes it easy to override/historize in git.

> ...and extension build (to create zip) use a different config :
> 
>     * https://github.com/georchestra/mapstore2-cadastrapp/blob/master/package.json#L113

now that's a good catch. with that local diff on top of this branch, and the `owners.svg` file in the toplevel `assets` subdir:
```
[27/12 10:15] landry@build.fluela:/data/src/georchestra/mapstore2-cadastrapp $git diff build/extension/prod-webpack.config.js 
diff --git a/build/extension/prod-webpack.config.js b/build/extension/prod-webpack.config.js
index bf61990..73431f8 100644
--- a/build/extension/prod-webpack.config.js
+++ b/build/extension/prod-webpack.config.js
@@ -12,6 +12,7 @@ const plugins = [
     new CopyPlugin([
         { from: path.resolve(__dirname, "..", "..", "assets", "translations"), to: "translations" },
+        { from: path.resolve(__dirname, "..", "..", "assets", "owners.svg"), to: "owners.svg" },
         { from: path.resolve(__dirname, "..", "..", "assets", "index.json"), to: "index.json" }

     ]),
     new ZipPlugin({
         filename: `${name}.zip`,

```
and building the extension:
```
[27/12 10:15] landry@build.fluela:/data/src/georchestra/mapstore2-cadastrapp $npm install && npm run ext:build
```
the resulting zip contains the svg file at the toplevel dir:
```
[27/12 10:22] landry@build.fluela:/data/src/georchestra/mapstore2-cadastrapp $unzip -l dist/Cadastrapp.zip |grep svg
     5551  2024-12-27 10:21   spritesheet.svg
   444379  2024-12-27 10:21   fontawesome-webfont.svg
      570  2024-12-27 10:21   symbolMissing.svg
     4322  2024-12-27 10:21   owners.svg
```
that could be used/reached from the extension code probably the same way how translations/index.json is imported/used. Or find the right magic so that the svg is copied under `assets/img` within the zip ?

btw i think the same thing/construct should be done to install the svg in #204 for the extension icon. instead of inlining the svg in a js file....

Gaetanbrl commented:
> the resulting zip contains the svg file at the toplevel dir:

Indeed, we have this behvior with SVG but it's a real ngithmare :

- in dev mode -> All works fine if we locate SVG file in js/extension/assets folder
- we start build
- SVG is at the top level and path is wrong

So, to avoid to disrupt webpack process and ease next / potential upgrade (if process changes) i'd suggest to use SVG string (juste copy / past SVG to constant file seems easy).

> Ou trouver la bonne magie pour que le svg soit copié sous assets/img dans le zip ?

That's the question we discussed this morning with @LPoin ...

LPoin committed:
modify plot search icon color to maintain the overall consistency of cadastrapp

LPoin commented:
About our SVG issue :
```
const fileLoader = {
    test: /\.(ttf|eot|svg)(\?v=[0-9].[0-9].[0-9])?$/,
    use: [{
        loader: 'file-loader',
        options: {
            name: "[name].[ext]"
        }
    }]
};
```
It treats svg as modules and places them in the defined output, in this case the root.

Tested solutions : 
- Add our custom icon in glyphIcons : need external app to include our custom svg, but depprecated (bootstrap 3 not maintained).
- Add external loader : don't work neither.
- Modify fileLoader in prod-webpack.config.js in build/extension (file used for npm run ext:build) to put svg in "asset/img/[name].[ext]". The path is not recognised once the build is complete.

Current test :
- Modify prod-webpack.config.js to create new paths according to relative path of SVG file imported.

LPoin commented:
I  try to modify prod-webpack.config.js (used in ext:build run) to fiw svg issue. 
Webpack is now in version 5.54.0 for 3 months and many elements don't works as before.
I modify fileLoader to 
```
const fileLoader = {
    test: /\.(ttf|eot|svg)(\?v=[0-9].[0-9].[0-9])?$/,
    type: 'asset/resource', // file loader for v5 webpack
    generator: {
        filename: 'assets/[name][ext]',
        publicPath: '/',
    },
};
```
as it work in V5.
I have to manage polyfills as webpack 5 removed automatics polyfills for node.js.
And I add alias to manage import with alias in files and supported extensions
```
alias: {
            "@mapstore/patcher": path.resolve(__dirname, '..', '..', "node_modules", "@mapstore", "patcher"),
            "@mapstore": path.resolve(__dirname, '..', '..', "MapStore2", "web", "client"),
            "@js": path.resolve(__dirname, '..', '..', "js"),
        },
extensions: ['.js', '.jsx', '.json']
```
With all these changes, build is fixed, svg are stored in the right place but I still have an error
![image](https://github.com/user-attachments/assets/e55ea309-246e-44e4-b432-7fc35f566cd9)

It might need modifications elsewhere. 
I'll create an issue on mapStore-Extension linked to this one.

LPoin commented:
An issue have been created in mapstore-extension [here](https://github.com/geosolutions-it/MapStoreExtension/issues/63) to solve the svg problem.
We could use the current solution in this PR and in [PR-204](https://github.com/georchestra/mapstore2-cadastrapp/pull/204) and I'll create a new issue to solve the problem after mapstore-extension solution.

pierrejego commented on js/extension/plugins/cadastrapp/toolbar/toolbarIcons.js:
```diff
@@ -0,0 +1,59 @@
+   id="svg5"
+   inkscape:version="1.1 (c68e22c387, 2021-05-23)"
+   sodipodi:docname="interface__cadastrApp.svg"
+   inkscape:export-filename="C:\Users\Agathe\Desktop\Cadastrapp\Refonte_interface\Cadastrapp__interfaceSearchProp.svg"
```
can it be removed from svg ?

pierrejego commented:
Conflict need to be merge

LPoin committed:
Merge branch 'master' into issue-199

LPoin committed:
clean svg code

LPoin committed:
clean code
Merge branch 'issue-199' of https://github.com/jdev-org/mapstore2-cadastrapp into issue-199

pierrejego commented:
Test OK

landryb commented:
the PR added `js/extension/assets/owners.svg` to the repo, which seems unused as inlined in `js/extension/plugins/cadastrapp/toolbar/toolbarIcons.js` - can it be removed to avoid confusion ?
